src/node_jsvmapi.cc

```diff
@@ -1569,122 +1600,153 @@ napi_status napi_coerce_to_string(napi_env e, napi_value v, napi_value* result)
+// Deletes a reference. The referenced value is released, and may be GC'd unless there
+// are other references to it.
+napi_status napi_delete_reference(napi_env e, napi_ref ref) {
   NAPI_PREAMBLE(e);
-  CHECK_ARG(result);
+  CHECK_ARG(ref);
+
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  delete reference;
 
-  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Persistent<v8::Value> *thePersistent =
-      new v8::Persistent<v8::Value>(
-          isolate, v8impl::V8LocalValueFromJsValue(v));
-  thePersistent->SetWeak(static_cast<int*>(nullptr), v8impl::WeakRefCallback,
-                         v8::WeakCallbackType::kParameter);
-  // need to mark independent?
-  *result = v8impl::JsWeakRefFromV8PersistentValue(thePersistent);
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_get_weakref_value(napi_env e, napi_weakref w, napi_value* result) {
+// Increments the reference count, optionally returning the resulting count. After this call the
+// reference will be a strong reference because its refcount is >0, and the referenced object is
+// effectively "pinned". Calling this when the refcount is 0 and the object is unavailable
+// results in an error.
+napi_status napi_reference_addref(napi_env e, napi_ref ref, int* result) {
   NAPI_PREAMBLE(e);
-  CHECK_ARG(result);
+  CHECK_ARG(ref);
 
-  v8::Isolate *isolate = v8impl::V8IsolateFromJsEnv(e);
-  v8::Persistent<v8::Value> *thePersistent =
-      v8impl::V8PersistentValueFromJsWeakRefValue(w);
-  v8::Local<v8::Value> v =
-      v8::Local<v8::Value>::New(isolate, *thePersistent);
-  if (v.IsEmpty()) {
-    *result = nullptr;
-    return GET_RETURN_STATUS();
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  int count = reference->AddRef();
+
+  if (result != nullptr) {
+    *result = count;
+  }
+
+  return GET_RETURN_STATUS();
+}
+
+// Decrements the reference count, optionally returning the resulting count. If the result is
+// 0 the reference is now weak and the object may be GC'd at any time if there are no other
+// references. Calling this when the refcount is already 0 results in an error.
+napi_status napi_reference_release(napi_env e, napi_ref ref, int* result) {
+  NAPI_PREAMBLE(e);
+  CHECK_ARG(ref);
+
+  v8impl::Reference* reference = reinterpret_cast<v8impl::Reference*>(ref);
+  int count = reference->Release();
+  if (count < 0) {
+    return napi_set_last_error(napi_generic_failure);
+  }
+
+  if (result != nullptr) {
+    *result = count;
   }
-  *result = v8impl::JsValueFromV8LocalValue(v);
+
   return GET_RETURN_STATUS();
 }
 
-napi_status napi_release_weakref(napi_env e, napi_weakref w) {
+// Attempts to get a referenced value. If the reference is weak, the value might no longer be
+// available, in that case the call is still successful but the result is NULL.
+napi_status napi_get_reference_value(napi_env e, napi_ref ref, napi_value* result) {
```

Comment:
I've wondered about this in the past; do you think it would be valuable to also have a bool out parameter to indicate success instead of returning nullptr for the result? Could it be the case that nullptr result is a valid (and indistinguishable from failure) value?